pkg/web/render.go

```diff
@@ -35,12 +38,16 @@ func Renderer(dir, leftDelim, rightDelim string) Middleware {
 	return func(next http.Handler) http.Handler {
 		return http.HandlerFunc(func(rw http.ResponseWriter, req *http.Request) {
 			ctx := FromContext(req.Context())
+			ctx.template = t
 			if Env == DEV {
-				if t, err = compileTemplates(fs, leftDelim, rightDelim); err != nil {
+				tt, err, _ := devEnvGr.Do("dev", func() (any, error) {
+					return compileTemplates(fs, leftDelim, rightDelim)
+				})
```

Comment:
I'm curious why you're using singleflight, rather than [`sync.OnceValues`](https://pkg.go.dev/sync#OnceValues) or another `sync.Once*` function?

You could also do:

```suggestion
				tt, err := sync.OnceValues(func() (*template.Template, error) {
					return compileTemplates(fs, leftDelim, rightDelim)
				})()
```

Reply:
The idea of `sync.OnceValues` is that you share the returned function, and then run it from all the goroutines you want. The function you pass to it will only run once, and all the goroutines that call the returned function will get the first value returned (and if more than one goroutine race to run it for the first time, only one will get to actually run the wrapped function and the other will wait until it finishes and return the same values).

So it's similar to `sync.Once` (actually a wrapper). Calling `sync.OnceValues(...)()` would unconditionally run the function you pass to it as a consequence, so it would be a nop. On the other hand, having the function returned by `sync.OnceValues` saved to a variable and run within the body of the decorator for every request would cause recompilation to happen at most once (if we ever serve a request), but the idea of the decorator is that recompilation happens every time the handler runs. This is far from efficient, but is already what was there and works well for development.

Reply:
Ah, I guess the intent is what I wasn't understanding. It seems that you _do_ want to re-render templates in dev mode (because the content on-disk could be different), which is why `sync.OnceValues` isn't desired here.

I see now that `singleflight` solves the race by ensuring only one of a number of concurrent in-flight requests will trigger rendering.

👍